repose-aggregator/tests/functional-tests/src/integrationTest/groovy/features/filters/apivalidator/ApiValidatorMultiTenantTest.groovy

```diff
@@ -0,0 +1,475 @@
+import static javax.servlet.http.HttpServletResponse.SC_OK
+import static org.openrepose.commons.utils.http.OpenStackServiceHeader.ROLES
+import static org.openrepose.commons.utils.http.PowerApiHeader.RELEVANT_ROLES
+import static org.openrepose.commons.utils.string.Base64Helper.base64EncodeUtf8
+
+class ApiValidatorMultiTenantTest extends ReposeValveTest {
+
+    static def mapHeaderValue = base64EncodeUtf8("""{
+      |   "1" : ["a:admin","foo","bar"],
+      |   "2" : ["a:creator", "foo", "a:observer"],
+      |   "3" : ["a:updater", "bar", "biz", "a:creator"],
+      |   "4" : ["a:observer"],
+      |   "5" : ["a:admin", "bar", "biz", "a:creator"],
+      |   "6" : ["biz", "baz"],
+      |   "7" : ["role with spaces", "biz"],
+      |   "8" : ["a:patcher"]
+      |}
+      |""".stripMargin())
+
+    static def observerRequests = [
+        ["GET", "/v1/resource"]
+    ]
+
+    static def creatorRequests = [
+        ["POST", "/v1/resource"],
+        ["POST", "/v1/resource/other"]
+    ]
+    static def updaterRequests = [
+        ["PUT", "/v1/resource"],
+        ["PUT", "/v1/resource/other"]
+    ]
+    static def adminOnlyRequests = [
+        ["DELETE", "/v1/resource"],
+        ["DELETE", "/v1/resource/other"]
+    ]
+
+
+    def setupSpec() {
+        deproxy = new Deproxy()
+        deproxy.addEndpoint(properties.targetPort)
+
+        def params = properties.getDefaultTemplateParams()
+        repose.configurationProvider.applyConfigs("common", params)
+        repose.configurationProvider.applyConfigs("features/filters/apivalidator/multitenant", params)
+        repose.start()
+
+        repose.waitForNon500FromUrl(reposeEndpoint)
+    }
+
+    @Unroll("Should succeed on GET /v1/resource/other when appropriate headers are set (#tenants)")
+    def "Should succeed when appropriate headers are set"() {
+        when:
+        MessageChain mc = deproxy.makeRequest(
+            url: "$reposeEndpoint/v1/resource/other",
+            method: "GET",
+            headers: [
+                "X-Tenant"   : tenants,
+                "X-Other"    : "other",
+                (ROLES)      : "foo,bar",
+                "X-Map-Roles": mapHeaderValue])
+
+        then:
+        mc.getReceivedResponse().getCode() as Integer == SC_OK
+        mc.handlings.size() == 1
+        List<String> relevantRolesValues = mc.handlings[0].request.headers.findAll(RELEVANT_ROLES)
+        relevantRolesValues.containsAll(["foo", "bar"])
+
+        where:
+        tenants << ["1,5", "4", "4,2", "1,2,3", "5,2", "1,3", "5,3", "1", "1,5,7", "1,5,6"]
+    }
+
+    @Unroll("Should succeed on #method #url when appropriate headers are set (1, 5)")
+    def "Should succeed when appropriate headers are set (1, 5)"() {
+        when:
+        MessageChain mc = deproxy.makeRequest(
+            url: "$reposeEndpoint$url",
+            method: method,
+            headers: [
+                "X-Tenant"   : "1,5",
```

Comment:
So this test shows that the required role must be present on tenant 1 OR 5, right? Rather than AND. Since we pass this filter, but only tenant 5 has role `a:creator` when we `POST` to `/v1/resource/other`.

Reply:
Nope. It's an `AND`. Both tenants `1` and `5` have the `a:admin` role, so we're good.